modules/mkdocs.py

```diff
@@ -0,0 +1,165 @@
+            site = sites[name]
+            site['aliases'].append(alias)
+
+        elif cmd in ['search', 'find']:
+            name_or_alias = args.pop(0).lower()
+            site = self.get_site(name_or_alias, room=room)
+            if not site:
+                return await bot.send_text(room, event, f'Sorry, could not find settings for site or alias `{name_or_alias}`')
+            results = self.search_index(site, args)
+            if len(results):
+                msg = ['I found these results:']
+                for (_, result) in results[:3]:
+                    msg.append(f'- [{result.title}]({site["url"]}/{result.location}')
+                # msg.append('For more results, see: ')
+                return await bot.send_text(room, event, '\n'.join(msg))
+            else:
+                return await bot.send_text(room, event, 'Unable to find a reference for the search.')
+
+    async def add_site(self, bot=None, room=None, event=None, args=[]):
+        name, url = args
+        sites = self.rooms[room.room_id]
+        if sites.get(name):
+            self.rooms[room.room_id][name]['url'] = url
+            await bot.send_text(room, event, f'Updated url for site `{name}` for searching')
+        else:
+            self.rooms[room.room_id][name] = {
+                'url': url,
+                'aliases': []
+            }
+            await bot.send_text(room, event, f'Added new site `{name}` for searching')
+        await self.populate_index(name, room=room)
+        bot.save_settings()
+
+    def get_site(self, name_or_alias, room=None):
+        sites = self.rooms[room.room_id]
+        if sites.get(name_or_alias):
+            return sites[name_or_alias]
+        else:
+            for site in sites:
+                if name_or_alias in site['alises']:
+                    return site
+        return None
+
+
+
+    async def search_index(self, site, keywords):
+        matches = []
+        for doc in reversed(site['docs']):
+            weight = 0
+            for keyword in keywords:
+                if re.search(keyword, doc['text'], re.IGNORECASE):
+                    weight = weight + 1
+            if weight > 0:
+                matches.append((weight, doc))
+        if len(keywords) > 1:
+            matches.sort(key=lambda tup: tup[0])
+        return matches
+
+    async def populate_index(self, name, room=None):
+        url = self.rooms[room.room_id][name]['url']
+        response = requests.get(f'{url}/search/search_index.json').json()
+        self.sites[name] = {
+            'config': response['config'],
+            'docs': []
+        }
+        for doc in response['docs']:
+            parts = doc.location.split('/')
+            if len(parts) < 2:
+                continue
+            *_, tail = parts
+            if tail.startswith("#"):
+                # MKDocs search has index for different headers. Ignore.
+                continue
+
+            self.sites[name]['docs'].append(doc)
```

Comment:
I recommend against storing the index in `self`. Every time a `bot.save_settings()` happens, `self.get_settings()` is called for every bot module and that data is pushed to the Matrix server. That means every site's search index is pushed up every time. That's a lot of data, I'm not sure whether there's a maximum we might hit for userdata in Synapse.

Besides, since the search index is already fetched from an existing site, that can be the source of truth we pull from instead of the Matrix server.

Reply:
Store it in a different object, and don't add this one to `self.get_settings()`
```suggestion
            self.sitedata[name].append(doc)
```

Reply:
Ahh, good point. Didn't know it got persisted to the synapse server, just thought it was storage local to the bot. Thanks for pointing that out.